File: java/src/test/java/tech/hearth/crypto/ApiKeyEnvelopeTest.java
package tech.hearth.crypto;

import static org.junit.jupiter.api.Assertions.assertArrayEquals;
import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertNull;
import static org.junit.jupiter.api.Assertions.assertThrows;
import static org.junit.jupiter.api.Assertions.assertTrue;

import java.time.Instant;
import java.util.Arrays;
import java.util.HashSet;
import java.util.Set;

import org.junit.jupiter.api.Test;
import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.EnumSource;

class ApiKeyEnvelopeTest {

    private static final Instant NOW = Instant.parse("2026-08-05T12:00:00Z");
    private static final Instant LATER = NOW.plusSeconds(3600);

    private static ApiKeyEnvelope.Metadata meta() {
        return ApiKeyEnvelope.Metadata.of("prod/ingest-api", LATER);
    }

    private static ApiKeyEnvelope.Opened open(X25519.Keypair recipient, byte[] envelope) {
        return ApiKeyEnvelope.open(recipient.secretKey(), envelope, NOW, Crypto.defaultBackend());
    }

    @ParameterizedTest
    @EnumSource(Hpke.Suite.class)
    void sealOpenRoundTrip(Hpke.Suite suite) {
        X25519.Keypair recipient = X25519.generateKeypair();
        char[] apiKey = ApiKeyEnvelope.randomApiKey();

        byte[] envelope = ApiKeyEnvelope.seal(
                recipient.publicKey(), apiKey, meta(), suite, Crypto.defaultBackend());

        // 20-byte fixed header + metadata + 32-byte enc + 32-byte key + 16-byte tag.
        int metadataLength = 1 + "prod/ingest-api".length() + Long.BYTES;
        assertEquals(20 + metadataLength + 32 + 48, envelope.length);

        ApiKeyEnvelope.Opened opened = open(recipient, envelope);
        assertArrayEquals(apiKey, opened.apiKey());
        assertEquals("prod/ingest-api", opened.metadata().keyId());
        assertEquals(LATER, opened.metadata().notAfter());

        opened.wipe();
        assertArrayEquals(new char[ApiKeyEnvelope.API_KEY_LENGTH], opened.apiKey());
    }

    @Test
    void headerCarriesTheSuiteAndRecipientFingerprint() {
        X25519.Keypair recipient = X25519.generateKeypair();
        byte[] envelope = ApiKeyEnvelope.seal(recipient.publicKey(), ApiKeyEnvelope.randomApiKey(), meta());

        assertEquals("HKE1", new String(Arrays.copyOf(envelope, 4), java.nio.charset.StandardCharsets.US_ASCII));
        assertEquals(Hpke.KEM_ID, be16(envelope, 4));
        assertEquals(Hpke.KDF_ID, be16(envelope, 6));
        assertEquals(ApiKeyEnvelope.DEFAULT_SUITE.aeadId(), be16(envelope, 8));
        assertArrayEquals(ApiKeyEnvelope.fingerprint(recipient.publicKey()),
                Arrays.copyOfRange(envelope, 10, 18));
    }

    @Test
    void rejectsEnvelopeForAnotherRecipient() {
        X25519.Keypair recipient = X25519.generateKeypair();
        X25519.Keypair other = X25519.generateKeypair();
        byte[] envelope = ApiKeyEnvelope.seal(recipient.publicKey(), ApiKeyEnvelope.randomApiKey(), meta());

        IllegalArgumentException e = assertThrows(IllegalArgumentException.class, () -> open(other, envelope));
        assertTrue(e.getMessage().contains("different recipient key"), e.getMessage());
    }

    @Test
    void rejectsExpiredEnvelope() {
        X25519.Keypair recipient = X25519.generateKeypair();
        byte[] envelope = ApiKeyEnvelope.seal(recipient.publicKey(), ApiKeyEnvelope.randomApiKey(),
                ApiKeyEnvelope.Metadata.of("short-lived", NOW.minusSeconds(1)));

        IllegalArgumentException e = assertThrows(IllegalArgumentException.class, () -> open(recipient, envelope));
        assertTrue(e.getMessage().startsWith("envelope expired"), e.getMessage());
    }

    @Test
    void metadataWithoutExpiryNeverExpires() {
        X25519.Keypair recipient = X25519.generateKeypair();
        byte[] envelope = ApiKeyEnvelope.seal(
                recipient.publicKey(), ApiKeyEnvelope.randomApiKey(), ApiKeyEnvelope.Metadata.of("forever"));

        ApiKeyEnvelope.Opened opened = ApiKeyEnvelope.open(
                recipient.secretKey(), envelope, Instant.parse("2099-01-01T00:00:00Z"), Crypto.defaultBackend());
        assertNull(opened.metadata().notAfter());
    }

    /** Every header byte is AAD, so relabelling the metadata breaks the tag. */
    @Test
    void rejectsTamperedMetadata() {
        X25519.Keypair recipient = X25519.generateKeypair();
        byte[] envelope = ApiKeyEnvelope.seal(recipient.publicKey(), ApiKeyEnvelope.randomApiKey(),
                ApiKeyEnvelope.Metadata.of("prod/ingest-api", LATER));

        // Flip the last byte of the expiry timestamp, still inside the header.
        byte[] tampered = envelope.clone();
        int metadataEnd = 20 + 1 + "prod/ingest-api".length() + Long.BYTES;
        tampered[metadataEnd - 1] ^= 0x01;

        IllegalArgumentException e = assertThrows(IllegalArgumentException.class, () -> open(recipient, tampered));
        assertTrue(e.getMessage().contains("not authentic"), e.getMessage());
    }

    @Test
    void rejectsTamperedCiphertextAndEncapsulatedKey() {
        X25519.Keypair recipient = X25519.generateKeypair();
        byte[] envelope = ApiKeyEnvelope.seal(recipient.publicKey(), ApiKeyEnvelope.randomApiKey(), meta());

        for (int offset : new int[] {envelope.length - 1, envelope.length - 40, envelope.length - 60}) {
            byte[] tampered = envelope.clone();
            tampered[offset] ^= 0x01;
            assertThrows(IllegalArgumentException.class, () -> open(recipient, tampered),
                    "flipping byte " + offset + " should not open");
        }
    }

    @Test
    void rejectsMalformedEnvelopes() {
        X25519.Keypair recipient = X25519.generateKeypair();
        byte[] envelope = ApiKeyEnvelope.seal(recipient.publicKey(), ApiKeyEnvelope.randomApiKey(), meta());

        assertThrows(IllegalArgumentException.class, () -> open(recipient, new byte[3]));

        byte[] wrongMagic = envelope.clone();
        wrongMagic[0] = 'X';
        assertThrows(IllegalArgumentException.class, () -> open(recipient, wrongMagic));

        byte[] truncated = Arrays.copyOf(envelope, envelope.length - 1);
        assertThrows(IllegalArgumentException.class, () -> open(recipient, truncated));

        byte[] unknownAead = envelope.clone();
        unknownAead[9] = (byte) 0xff;
        assertThrows(IllegalArgumentException.class, () -> open(recipient, unknownAead));
    }

    @Test
    void rejectsApiKeysOfTheWrongShape() {
        byte[] publicKey = X25519.generateKeypair().publicKey();
        assertThrows(IllegalArgumentException.class,
                () -> ApiKeyEnvelope.seal(publicKey, "tooshort".toCharArray(), meta()));
        assertThrows(IllegalArgumentException.class,
                () -> ApiKeyEnvelope.seal(publicKey, "0123456789012345678901234567890!".toCharArray(), meta()));
    }

    @Test
    void metadataRejectsEmptyOrOversizedKeyId() {
        assertThrows(IllegalArgumentException.class, () -> ApiKeyEnvelope.Metadata.of(""));
        assertThrows(IllegalArgumentException.class, () -> ApiKeyEnvelope.Metadata.of("k".repeat(256)));
    }

    @Test
    void randomApiKeyIsAlphanumericAndCoversTheAlphabet() {
        Set<Character> seen = new HashSet<>();
        for (int i = 0; i < 200; i++) {
            char[] key = ApiKeyEnvelope.randomApiKey();
            assertEquals(ApiKeyEnvelope.API_KEY_LENGTH, key.length);
            for (char c : key) {
                assertTrue(Character.isLetterOrDigit(c) && c < 128, "not ASCII alphanumeric: " + c);
                seen.add(c);
            }
        }
        // 6400 draws over a 62-character alphabet: every character should appear.
        assertEquals(62, seen.size());
    }

    private static int be16(byte[] bytes, int offset) {
        return ((bytes[offset] & 0xff) << 8) | (bytes[offset + 1] & 0xff);
    }
}

File: typescript/test/apikeyenvelope.test.ts
import assert from "node:assert/strict";
import test from "node:test";
import { apikeyenvelope as env, hpke, x25519 } from "../src/index.ts";

const NOW = new Date("2026-08-05T12:00:00Z");
const LATER = new Date(NOW.getTime() + 3600 * 1000);

const meta = (): env.Metadata => env.metadata("prod/ingest-api", LATER);

const ALL_SUITES: hpke.Suite[] = [
  "X25519_SHA256_AES128GCM",
  "X25519_SHA256_AES256GCM",
  "X25519_SHA256_CHACHA20POLY1305",
];

test("seal/open round trip on every suite", () => {
  for (const suite of ALL_SUITES) {
    const recipient = x25519.generateKeypair();
    const apiKey = env.randomApiKey();
    const apiKeyCopy = apiKey.slice();

    const envelope = env.seal(recipient.publicKey, apiKey, meta(), suite);

    // 20-byte fixed header + metadata + 32-byte enc + 32-byte key + 16-byte tag.
    const metadataLength = 1 + "prod/ingest-api".length + 8;
    assert.equal(envelope.length, 20 + metadataLength + 32 + 48);

    const opened = env.open(recipient.secretKey, envelope, NOW);
    assert.deepEqual(opened.apiKey, apiKeyCopy);
    assert.equal(opened.metadata.keyId, "prod/ingest-api");
    assert.equal(opened.metadata.notAfter?.getTime(), LATER.getTime());

    env.wipe(opened);
    assert.deepEqual(opened.apiKey, new Uint8Array(env.API_KEY_LENGTH));
  }
});

test("header carries the suite and recipient fingerprint", () => {
  const recipient = x25519.generateKeypair();
  const envelope = env.seal(recipient.publicKey, env.randomApiKey(), meta());

  assert.equal(new TextDecoder().decode(envelope.slice(0, 4)), "HKE1");
  assert.equal(be16(envelope, 4), hpke.KEM_ID);
  assert.equal(be16(envelope, 6), hpke.KDF_ID);
  assert.equal(be16(envelope, 8), hpke.aeadId(env.DEFAULT_SUITE));
  assert.deepEqual(envelope.slice(10, 18), env.fingerprint(recipient.publicKey));
});

test("rejects envelope for another recipient", () => {
  const recipient = x25519.generateKeypair();
  const other = x25519.generateKeypair();
  const envelope = env.seal(recipient.publicKey, env.randomApiKey(), meta());

  assert.throws(() => env.open(other.secretKey, envelope, NOW), /different recipient key/);
});

test("rejects expired envelope", () => {
  const recipient = x25519.generateKeypair();
  const envelope = env.seal(
    recipient.publicKey,
    env.randomApiKey(),
    env.metadata("short-lived", new Date(NOW.getTime() - 1000)),
  );

  assert.throws(() => env.open(recipient.secretKey, envelope, NOW), /envelope expired/);
});

test("metadata without expiry never expires", () => {
  const recipient = x25519.generateKeypair();
  const envelope = env.seal(recipient.publicKey, env.randomApiKey(), env.metadata("forever"));

  const opened = env.open(recipient.secretKey, envelope, new Date("2099-01-01T00:00:00Z"));
  assert.equal(opened.metadata.notAfter, null);
});

test("rejects tampered metadata — every header byte is AAD", () => {
  const recipient = x25519.generateKeypair();
  const envelope = env.seal(recipient.publicKey, env.randomApiKey(), env.metadata("prod/ingest-api", LATER));

  const tampered = envelope.slice();
  const metadataEnd = 20 + 1 + "prod/ingest-api".length + 8;
  tampered[metadataEnd - 1] ^= 0x01;

  assert.throws(() => env.open(recipient.secretKey, tampered, NOW), /not authentic/);
});

test("rejects tampered ciphertext and encapsulated key", () => {
  const recipient = x25519.generateKeypair();
  const envelope = env.seal(recipient.publicKey, env.randomApiKey(), meta());

  for (const offset of [envelope.length - 1, envelope.length - 40, envelope.length - 60]) {
    const tampered = envelope.slice();
    tampered[offset] ^= 0x01;
    assert.throws(() => env.open(recipient.secretKey, tampered, NOW), `flipping byte ${offset} should not open`);
  }
});

test("rejects malformed envelopes", () => {
  const recipient = x25519.generateKeypair();
  const envelope = env.seal(recipient.publicKey, env.randomApiKey(), meta());

  assert.throws(() => env.open(recipient.secretKey, new Uint8Array(3), NOW));

  const wrongMagic = envelope.slice();
  wrongMagic[0] = "X".charCodeAt(0);
  assert.throws(() => env.open(recipient.secretKey, wrongMagic, NOW));

  const truncated = envelope.slice(0, envelope.length - 1);
  assert.throws(() => env.open(recipient.secretKey, truncated, NOW));

  const unknownAead = envelope.slice();
  unknownAead[9] = 0xff;
  assert.throws(() => env.open(recipient.secretKey, unknownAead, NOW));
});

test("rejects API keys of the wrong shape", () => {
  const publicKey = x25519.generateKeypair().publicKey;
  assert.throws(() => env.seal(publicKey, new TextEncoder().encode("tooshort"), meta()));
  assert.throws(() => env.seal(publicKey, new TextEncoder().encode("0123456789012345678901234567890!"), meta()));
});

test("metadata rejects empty or oversized key id", () => {
  assert.throws(() => env.metadata(""));
  assert.throws(() => env.metadata("k".repeat(256)));
});

test("randomApiKey is alphanumeric and covers the alphabet", () => {
  const seen = new Set<string>();
  for (let i = 0; i < 200; i++) {
    const key = env.randomApiKey();
    assert.equal(key.length, env.API_KEY_LENGTH);
    for (const b of key) {
      const c = String.fromCharCode(b);
      assert.ok(/[A-Za-z0-9]/.test(c), `not ASCII alphanumeric: ${c}`);
      seen.add(c);
    }
  }
  // 6400 draws over a 62-character alphabet: every character should appear.
  assert.equal(seen.size, 62);
});

function be16(bytes: Uint8Array, offset: number): number {
  return (bytes[offset] << 8) | bytes[offset + 1];
}
